storey/flow.py

```diff
@@ -320,21 +364,90 @@ def _user_fn_output_to_event(self, event, fn_result):
                 mapped_event.body = fn_result
             return mapped_event
 
-    def _check_step_in_flow(self, type_to_check):
+    def _check_step_in_flow(self, type_to_check, visited=None):
+        # initialize the visited set once at the top
+        if visited is None:
+            visited = set()
+
+        # detect cycles
+        if self in visited:
+            return False
+        visited.add(self)
+
+        # check this node
         if isinstance(self, type_to_check):
             return True
+
+        # check outlets
         for outlet in self._outlets:
-            if outlet._check_step_in_flow(type_to_check):
+            if outlet._check_step_in_flow(type_to_check, visited):
                 return True
+
+        # check recovery step
         if isinstance(self._recovery_step, Flow):
-            if self._recovery_step._check_step_in_flow(type_to_check):
+            if self._recovery_step._check_step_in_flow(type_to_check, visited):
                 return True
+
         elif isinstance(self._recovery_step, dict):
             for step in self._recovery_step.values():
-                if step._check_step_in_flow(type_to_check):
+                if step._check_step_in_flow(type_to_check, visited):
                     return True
         return False
 
+    def check_and_update_iteration_number(self, event) -> Optional[Callable]:
+        if hasattr(event, "_cyclic_counter") and self._max_iteration is not None:
+            counter = self.get_iteration_counter(event)
+            if counter >= self._max_iteration:
+                raise RuntimeError(f"Max iterations exceeded in step '{self.name}' for event {event.id}")
+            event._cyclic_counter[self.name] = counter + 1
+        else:
+            event._cyclic_counter = {self.name: 1}
+
+    def get_iteration_counter(self, event):
+        if not hasattr(event, "_cyclic_counter"):
+            return 0
+        else:
+            return event._cyclic_counter.get(self.name, 0)
+
+    def select_outlets(self, event) -> Optional[Collection[str]]:
+        """
+        Override this method to route events based on a custom logic. The default implementation will route all
+        events to all outlets.
+        """
+        return None
+
+    def set_next_outlets(self, outlet_names: Union[str, list[str]]):
+        """
+        Set the next outlets to which the event will be sent. This method can be used in conjunction with
+        select_outlets() to dynamically determine the outlets for the next event.
+
+        :param outlet_names: A collection of outlet names to which the next event should be sent.
+        """
+        if not self._name_to_outlet:
+            self._init_name_to_outlet()
+        self._selected_outlet = outlet_names if isinstance(outlet_names, list) else [outlet_names]
```

Comment:
Using an attribute like this isn't actually safe. It assumes that each step handles events one by one, which isn't necessarily the case. It's true of `Map`, not true of something like concurrent execution, parallel execution, batch, targets, etc... If this facility is only relevant to certain types of steps, the functionality could go into a new intermediate class in the hierarchy (if needed).